truffle/src/com.oracle.truffle.api.strings.test/src/com/oracle/truffle/api/strings/test/taint/TaintPropagationConcatenationTests.java

```diff
@@ -0,0 +1,118 @@
+    }
+
+    @Test
+    public void bothTaintedConcat() {
+        AbstractTruffleString a = taint("foo", 1);
+        AbstractTruffleString b = taint("bar", 2);
+        AbstractTruffleString c = concat(a, b);
+
+        assertTrue("TS should be tainted", isTainted(c));
+
+        Object[] taintArr = getTaint(c);
+        assertEquals("Taint should have combined length", length(a) + length(b), taintArr.length);
+        assertArrayEquals("Arrays should be concatenated", new Object[] {1, 1, 1, 2, 2, 2}, taintArr);
+    }
+
+    @Test
+    public void aTaintedConcat() {
+        AbstractTruffleString a = taint("foo", 1);
+        AbstractTruffleString b = from("bar");
+        AbstractTruffleString c = concat(a, b);
+
+        assertTrue("TS should be tainted", isTainted(c));
+
+        Object[] taintArr = getTaint(c);
+        assertEquals("Taint should have combined length", length(a) + length(b), taintArr.length);
+        assertArrayEquals("Arrays should be concatenated", new Object[] {1, 1, 1, null, null, null}, taintArr);
+    }
+
+    @Test
+    public void bTaintedConcat() {
+        AbstractTruffleString a = from("foo");
+        AbstractTruffleString b = taint("bar", 2);
+        AbstractTruffleString c = concat(a, b);
+
+        assertTrue("TS should be tainted", isTainted(c));
+
+        Object[] taintArr = getTaint(c);
+        assertEquals("Taint should have combined length", length(a) + length(b), taintArr.length);
+        assertArrayEquals("Arrays should be concatenated", new Object[] {null, null, null, 2, 2, 2}, taintArr);
+    }
+
+    @Test
+    public void aEmpty() {
+        AbstractTruffleString a = from("");
+        AbstractTruffleString b = taint("bar", 2);
+        AbstractTruffleString c = concat(a, b);
+
+        assertTrue("TS should be tainted", isTainted(c));
+
+        Object[] taintArr = getTaint(c);
+        assertEquals("Taint should have combined length", length(a) + length(b), taintArr.length);
+        assertArrayEquals("Arrays should be concatenated", new Object[] {2, 2, 2}, taintArr);
+    }
+
+    @Test
+    public void bEmpty() {
+        AbstractTruffleString a = taint("foo", 1);
+        AbstractTruffleString b = from("");
+        AbstractTruffleString c = concat(a, b);
+
+        assertTrue("TS should be tainted", isTainted(c));
+
+        Object[] taintArr = getTaint(c);
+        assertEquals("Taint should have combined length", length(a) + length(b), taintArr.length);
+        assertArrayEquals("Arrays should be concatenated", new Object[] {1, 1, 1}, taintArr);
+    }
+
+    @Test
+    public void bothEmpty() {
+        AbstractTruffleString a = from("");
+        AbstractTruffleString b = from("");
+        AbstractTruffleString c = concat(a, b);
+
+        assertFalse("Empty TS cannot be tainted", isTainted(c));
+
+        Object[] taintArr = getTaint(c);
+        assertNull(taintArr);
+    }
+
+    @Test
+    public void lazyConcat() {
```

Comment:
I haven't tackled lazy string concatenation as of now

Reply:
Ok, not sure that's even needed. Just always concatenating eagerly is fine.

Reply:
Well, I tried to make lazy concatenations also work now 🤷